Mathlib/Combinatorics/SimpleGraph/Tutte.lean

```diff
@@ -0,0 +1,120 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pim Otte
+-/
+
+import Mathlib.Combinatorics.SimpleGraph.Matching
+import Mathlib.Combinatorics.SimpleGraph.Operations
+import Mathlib.Combinatorics.SimpleGraph.UniversalVerts
+import Mathlib.Data.Fintype.Card
+
+/-!
+# Tutte's theorem (work in progress)
+
+## Main definitions
+* `SimpleGraph.TutteViolator G u` is a set of vertices `u` which certifies non-existence of a
+  perfect matching.
+-/
+
+namespace SimpleGraph
+
+universe u
+variable {V : Type u} {G G' : SimpleGraph V} {u x v' w : V} [Fintype V]
+
+/-- A set certifying non-existence of a perfect matching. -/
+def IsTutteViolator (G : SimpleGraph V) (u : Set V) : Prop :=
+  u.ncard < ((⊤ : G.Subgraph).deleteVerts u).coe.oddComponents.ncard
+
+/-- Given a graph in which the universal vertices do not violate Tutte's condition,
+if the graph decomposes into cliques, there exists a matching that covers
+everything except some universal vertices.
+
+This lemma is marked private, because
+it is strictly weaker than `IsPerfectMatching.exists_of_isClique_supp` -/
+private lemma Subgraph.IsMatching.exists_verts_compl_subset_universalVerts
+    (h : ¬IsTutteViolator G G.universalVerts)
+    (h' : ∀ (K : G.deleteUniversalVerts.coe.ConnectedComponent),
+    G.deleteUniversalVerts.coe.IsClique K.supp) :
+    ∃ M : Subgraph G, M.IsMatching ∧ M.vertsᶜ ⊆ G.universalVerts := by
+  classical
+  have hrep := ConnectedComponent.Represents.image_out
+      G.deleteUniversalVerts.coe.oddComponents
+  -- First we match one node from each odd component to a universal vertex
+  obtain ⟨t, ht, M1, hM1⟩ := Subgraph.IsMatching.exists_of_universalVerts
+      (disjoint_image_val_universalVerts _).symm (by
+        simp only [IsTutteViolator, not_lt] at h
+        rwa [Set.ncard_image_of_injective _ Subtype.val_injective, hrep.ncard_eq])
+  -- Then we match all other nodes in components internally
+  have compMatching (K : G.deleteUniversalVerts.coe.ConnectedComponent) :
+      ∃ M : Subgraph G, M.verts = Subtype.val '' K.supp \ M1.verts ∧ M.IsMatching := by
+    have : G.IsClique (Subtype.val '' K.supp \ M1.verts) :=
+      ((h' K).of_induce).subset Set.diff_subset
+    rw [← this.even_iff_exists_isMatching (Set.toFinite _), hM1.1]
+    exact even_ncard_image_val_supp_sdiff_image_val_rep_union _ ht hrep
+  let M2 : Subgraph G := ⨆ (K : G.deleteUniversalVerts.coe.ConnectedComponent),
+    (compMatching K).choose
+  have hM2 : M2.IsMatching := by
+    apply Subgraph.IsMatching.iSup (fun c ↦ (compMatching c).choose_spec.2) (fun i j hij ↦ ?_)
+    rw [(compMatching i).choose_spec.2.support_eq_verts,
+        (compMatching j).choose_spec.2.support_eq_verts,
+        (compMatching i).choose_spec.1, (compMatching j).choose_spec.1]
+    exact Set.disjoint_of_subset Set.diff_subset Set.diff_subset <|
+      Set.disjoint_image_of_injective Subtype.val_injective <|
+        SimpleGraph.pairwise_disjoint_supp_connectedComponent _ hij
```

Comment:
Since you use `.choose` and `choose_spec` here a lot, this might be nicer if you do `choose a b c using compMatching`, or similar! (If you have already tried this and it's worse, then disregard this, but I thought it'd be worth mentioning in case you weren't already aware)

Reply:
I wasn't aware of this construction:)

I tried to refactor it, but I didn't really get anywhere. The issue here is the split between the definition of M2 and the proofs about it. I'm not really sure how I could unify those or make use of the `choose ... using` construct in the proofs while keeping the current definition of M2.

Reply:
See the last commit for how to do so 😉